Platform Tags

phase opened the pull request:
Add SpongeForge, SpongeVanilla, SpongeCommon, and Lantern tags. Sponge & Forge tags won't be shown if an SF/SV dependency is present. Read the comments below for more details.

Closes #643

phase committed:
SpongeForge & SpongeVanilla Tags

Closes #643

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

stephan-gh commented on app/ore/project/Dependency.scala:
```diff
@@ -40,4 +40,8 @@ object Dependency {
     */
   val ForgeId: String = "forge"
 
+  val SpongeForgeId: String = "spongeforge"
```
Problem: There is no `spongeforge`/`spongevanilla` plugin ID. Both implementations use `sponge` as plugin ID currently.

phase commented:
IDs:
* SpongeAPI - `spongeapi`
* SpongeCommon - `sponge`
* SpongeForge - `spongeforge`
* SpongeVanilla - `spongevanilla`

These will be getting Plugin Containers soon(tm).

XakepSDK commented:
@phase 
Why SpongeCommon - `sponge`?

phase commented:
Sponge uses `sponge` for a lot of things internally and we want something to have that mod id.

XakepSDK commented:
Sponge may have multiple implementations, saying that SpongeCommon is `sponge` is wrong.

Zidane commented:
@XakepSDK 

Doesn't matter, my implementation in common is always known as `sponge`.

All other ones are known differently. SpongeForge and SpongeVanilla will gain a container with their own id.

phase committed:
Abstract Platforms to PlatformCategories for future Platforms

If/when we want to add Bukkit/Spigot/Paper or whatever else, adding a
new platform is super simple.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase commented:
Here's a preview of a project creation with the new platform:
![2018-08-16-142248_1920x1080_scrot](https://user-images.githubusercontent.com/7685133/44236827-33effc00-a163-11e8-8d99-7c8d01df0859.png)

And here's a version list with different variations:

![2018-08-16-142308_1920x1080_scrot](https://user-images.githubusercontent.com/7685133/44236828-33effc00-a163-11e8-84e6-2dd00bb6f837.png)

The dependencies for each version:
* **1.0.4**: `forge@blah, spongeforge@blah, spongeapi@blah`
* **1.0.3**: `forge@blah, spongeapi@blah`
* **1.0.2**: `spongeforge@blah, spongeapi@blah`
* ~~1.0.1: ignore that one I messed up~~
* **1.0.0**: `spongeapi@blah`

SpongeCommon isn't tracked but I guess it could if we wanted it to be.
I rewrote the Platform system so that adding a new one is _super_ simple.

I'm also open to suggestions for the **COLORS OF TAGS!!!!!**, the most important part. Currently, SpongeForge and SpongeVanilla (which also works with this system) use the Sponge tag, but we can use anything.

phase committed:
Change SpongeVanilla Priority and Tag Colors

Multiple Platforms can now have the same Priority so that both are
shown.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase commented:
![colors!](https://cdn.discordapp.com/attachments/468560985171427329/479790087056523284/2018-08-16-161334_1920x1080_scrot.png)

SF & SV will both be displayed if they are both depended on (Platforms with the same priority are shown together).

phase committed:
SpongeCommon & Lantern, removed Dependency constants

The constants weren't needed.

Also *someone* used tabs. Those were removed.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase commented:
Colors have been changed!

![](https://cdn.discordapp.com/attachments/468560985171427329/479804962386345984/2018-08-16-171235_1920x1080_scrot.png)

And if a project's recommended version has more than 2 tags, the data will be excluded on the home page (this screenshot doesn't have the new colors).

![](https://cdn.discordapp.com/attachments/417090248279851028/479800033332690994/2018-08-16-165323_1920x1080_scrot.png)

phase commented:
working on a platform filter now

phase committed:
Working on Platform Filter

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

Katrix commented on app/controllers/Application.scala:
```diff
@@ -87,17 +85,21 @@ final class Application @Inject()(data: DataHelper,
+
+    // get the categories being queried
+    val categoryPlatformNames: List[String] = pcat.toList.flatMap(_.getPlatforms.map(_.name))
+    val platformNames: List[String] = pform.map(_.name).toList.:::(categoryPlatformNames).map(_.toLowerCase)
```
`::: ` should be called infix

Katrix commented on app/models/project/Project.scala:
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
+  /**
+    * Get a collection of tags that represent a project through its versions
+    */
+  def tags(implicit ec: ExecutionContext, service: ModelService = null): Future[List[Tag]] = {
```
Can we please try to get away from this pattern with an implicit value with a default null?

phase replied:
I just copied the code from `Version#tags`

Katrix commented on app/models/project/Project.scala:
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
+      val tagIds = versions.flatMap(_.tagIds).toSet.toList
+      // get all the tags for all the versions
+      this.service.access(classOf[Tag]).filter(t => t.id inSet tagIds).map { list =>
+        list.toSet.toList
```
There's a method for this called `distinct`

Katrix commented on app/models/project/Project.scala:
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
+      this.service.access(classOf[Tag]).filter(t => t.id inSet tagIds).map { list =>
+        list.toSet.toList
+          // get the latest tag from the versions
+          .groupBy[String](_.name)
```
Is the type parameter needed here?

Katrix commented on app/models/project/Project.scala:
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
+        list.toSet.toList
+          // get the latest tag from the versions
+          .groupBy[String](_.name)
+          .map(pair => {
```
Only use curly braces here. Also, name the pair's elements with a `case`

Katrix commented on app/models/project/Project.scala:
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
+                .filter(_.tagIds.contains(t.id.get))
+                .filter(!_.isDeleted)
+                // get the latest version
+                .map(_.createdAt.getOrElse(Timestamp.from(Instant.MIN)).toInstant.toEpochMilli)
```
Don't be afraid to get here. It's an error if `createdAt` is empty.

Katrix commented on app/models/project/Tag.scala:
```diff
@@ -23,22 +26,47 @@ case class Tag(override val id: Option[Int] = None,
+    */
+  def getFilledTag(service: ModelService)(implicit ex: ExecutionContext): Future[Tag] = {
+    val access = service.access(classOf[Tag])
+    for {
```
Just a styling preference, can we get rid of the curly braces for tag and yield here?

Katrix commented on app/ore/Platforms.scala:
```diff
@@ -1,13 +1,86 @@
+  def getPlatforms(dependencyIds: List[String]): List[Platform] = {
+    Platforms.values
+      .filter(p => dependencyIds.contains(p.dependencyId))
+      .groupBy[PlatformCategory](p => p.platformCategory)
```
Is the type parameter needed here?

Katrix commented on app/ore/Platforms.scala:
```diff
@@ -1,13 +1,86 @@
+  def getPlatformGhostTags(dependencies: List[Dependency]): List[Tag] = {
+    Platforms.values
+      .filter(p => dependencies.map(_.pluginId).contains(p.dependencyId))
+      .groupBy[PlatformCategory](p => p.platformCategory)
```
Is the type parameter needed here?

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -374,44 +370,23 @@ trait ProjectFactory {
+    for {
+      tags <- Future.sequence(Platforms.getPlatformGhostTags(version.dependencies).map(tag => tag.getFilledTag(service)))
+    } yield {
+      tags.map(tag => {
```
Only curly braces here

phase committed:
Platform Filter Highlight & Fix Project Tags

lots of code was cleaned up

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

Katrix commented on app/models/project/Project.scala:
```diff
@@ -507,27 +507,27 @@ case class Project(override val id: Option[Int] = None,
-                .filter(_.tagIds.contains(t.id.get))
+          .groupBy(_.name)
+          .map { case (_, tags) =>
+            tags.maxBy(
```
Curly braces here

phase committed:
Show Platforms in Dependencies Section

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase commented:
I updated the Dependencies section for versions. They now include links to the download pages for whatever the platform is.

![2018-08-18-145748_1920x1080_scrot](https://user-images.githubusercontent.com/7685133/44303516-469b3a00-a2f7-11e8-9ddf-da289c7dd5c8.png)

Here's a final view of the hompage being filtered by forge mods.

![2018-08-18-145606_1920x1080_scrot](https://user-images.githubusercontent.com/7685133/44303517-469b3a00-a2f7-11e8-82c7-42488c56e489.png)

phase committed:
Use a filtered Platforms list for Version Dependencies

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Start migration to PluginFileData

Instead of only using the mcmod.info to parse, we want a variety of
different files to get metadata from. This system rips out to old
reliance on Sponge's mcmod.info parser and creates a modular system
for adding metadata files.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Add Project Tags from PluginFileData

When a pending version is processed, we add specialized tags from the
PluginFileData.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Add Mixin Tag

Finally got this working. It reads the MANIFEST.MF

Closes #247

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

Katrix commented on app/models/project/Project.scala:
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
+    schema(service)
+    // get all the versions for the project
+    this.service.access(classOf[Version]).filter(_.projectId === id.get).flatMap { versions =>
+      val tagIds = versions.flatMap(_.tagIds).toSet.toList
```
Use distinct here

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -374,44 +370,22 @@ trait ProjectFactory {
-          tag
-        }
+  private def addDependencyTags(version: Version)(implicit ec: ExecutionContext): Future[Seq[ProjectTag]] = {
+    for {
```
Do we need the for comprehension here?

phase committed:
Code cleanup in Platforms and other places

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Put a limit to tag data length

We don't want people spamming the front page with giant tags.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Filter invalid dependency versions out

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

Katrix commented on app/ore/Platforms.scala:
```diff
@@ -1,13 +1,98 @@
+      .groupBy(_.platformCategory)
+      .flatMap(_._2.groupBy(_.priority).maxBy(_._1)._2)
+      .map(_.asInstanceOf[Platform])
+      .toList
```
Be explicit about the return type if you're explicit about the Seq implementation.

Katrix commented on app/ore/Platforms.scala:
```diff
@@ -1,13 +1,98 @@
+      .groupBy(_.platformCategory)
+      .flatMap(_._2.groupBy(_.priority).maxBy(_._1)._2)
+      .map(p => p.toGhostTag(dependencies.find(_.pluginId == p.dependencyId).get.version))
+      .toList
```
Be explicit about the return type if you're explicit about the Seq implementation.

Katrix commented on app/ore/Platforms.scala:
```diff
@@ -1,13 +1,98 @@
+  val name: String
+
+  def getPlatforms: Seq[Platforms.Value] = {
+    Platforms.values.filter(p => p.platformCategory == this).toList
```
Be explicit about the return type if you're explicit about the Seq implementation.

phase committed:
Merge branch 'feature/sf-tags' into feature/mixin-tag

phase committed:
Clean up code in PluginFileData

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Switch back to using plugin-meta

We're going to keep mcmod.info parsing in there.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Rename some methods in PluginFileData

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Change toList to toSeq

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase committed:
Merge pull request #646 from SpongePowered/feature/mixin-tag

Rewrite Metadata Ripper to support multiple types of files, and add Mixin Tag

phase commented:
I changed snyk not to check OreTestPlugin on PRs.
